### PR TITLE
feature: Single Asset Vault

### PR DESCRIPTION
**Pull request by thejohnfreeman**
Starting a draft PR for incremental review.

Based on:
- MPT branch with variant `STIssue`
- #5122 
- #5121

### CODE REVIEW

**Comment by Bronek** [comment]
> Is there an XLS document for this new feature? Thanks.

Hi Donovan, please note this is a Draft, meaning the implementation here should be in no way considered complete or reflecting the actual product which will be subject to amendment voting. Having said that, this project started with the discussion at https://github.com/XRPLF/XRPL-Standards/discussions/190

**Comment by Bronek** [comment]
@donovanhide , I just realized that I pointed you to the wrong specification, sorry about that. This PR implements [Single Asset Vault specification](https://github.com/XRPLF/XRPL-Standards/discussions/192). It is in relation to [Lending Protocol](https://github.com/XRPLF/XRPL-Standards/discussions/190) but a separate specification, because these two are related, but not  coupled. Single Asset Vault is a prerequisite for lending protocol(s).

File: src/test/app/Vault_test.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        // TODO: VaultSet (update) fail: tfPrivate flag
+        // TODO: VaultSet (update) fail: tfShareNonTransferable flag
+        // TODO: Payment to VaultSet.PA fail
+        // TODO: VaultDelete succeed
+        // TODO: VaultDelete fail: missing vault
+        // TODO: VaultSet (update) fail: missing vault
+
+        // TODO: VaultSet (create) fail: Asset is MPT but !CanTransfer
+        // TODO: VaultSet (create) fail: Asset is MPT but Locked
+
```
Comment (Bronek) [inline src/test/app/Vault_test.cpp:80]: really nice outline of planned tests here ! 👏

File: src/xrpld/app/tx/detail/AMMCreate.cpp [hunk 2]
```diff
@@ -214,25 +214,25 @@ applyCreate(
     auto const ammKeylet = keylet::amm(amount.issue(), amount2.issue());
 
     // Mitigate same account exists possibility
-    auto const ammAccount = createPseudoAccount(sb, ammKeylet.key);
+    auto const maybe = createPseudoAccount(sb, ammKeylet.key);
```
Comment (Bronek) [inline src/xrpld/app/tx/detail/AMMCreate.cpp:217]: how about `maybeAccount` ?

File: src/xrpld/app/tx/detail/VaultSet.cpp [hunk 3]
```diff
@@ -82,20 +85,36 @@ VaultSet::doApply()
...
-        if (auto ter = dirLink(view(), tx[sfAccount], vault); ter)
+        if (auto ter = dirLink(view(), owner, vault); ter)
             return ter;
-        auto pseudo = createPseudoAccount(view(), vault->key());
-        if (!pseudo)
-            return pseudo.error();
-        (*vault)[sfSequence] = tx[sfSequence];
-        (*vault)[sfOwner] = tx[sfAccount];
-        (*vault)[sfAccount] = pseudo.value();
+        auto maybe = createPseudoAccount(view(), vault->key());
```
Comment (Bronek) [inline src/xrpld/app/tx/detail/VaultSet.cpp:107]: how about `maybeAccount` ?

File: src/xrpld/app/tx/detail/VaultSet.cpp [hunk 4]
```diff
@@ -82,20 +85,36 @@ VaultSet::doApply()
...
+        if (!maybe)
+            return maybe.error();
+        auto& pseudo = *maybe;
+        auto pseudoId = (*pseudo)[sfAccount];
+        auto maybe2 = MPTokenIssuanceCreate::create(
```
Comment (Bronek) [inline src/xrpld/app/tx/detail/VaultSet.cpp:119]: how about `maybeMPT` ?

**Comment by thejohnfreeman** [comment]
Will open a new PR, easier to review, in my personal fork.

**Pull request closed without merging**